### ISSUE TITLE
\Symbols with parameters entered without parameters don't appear when typed

### ISSUE THREAD

**Issue by jhiswin** [arnog/mathlive#459]
\Symbols with parameters can't be entered without parameters, and the hinting also appears to not be working.  Some examples:
\frac 
\sqrt
\tilde
Hinting tooltip does not show a picture preview while typing
Expecting it to show the fraction/sqrt with placeholders after typeing \frac and pressing enter

Works in 0.35.1
Stopped working after 0.50.1

Chrome 81
Windows 10

**Comment by arnog** [comment]
That's an unintended side effect of the error checking that has been added recently.

That said, the Command Mode (i.e. entering a Latex fragment with "\") is on the roadmap to be deprecated (see #270). Is there a use case that would make preserving the Command Mode or fixing this worthwhile? There should be keyboard alternatives available for these (e.g. "/" for `\frac`, alt+v for `\sqrt`, etc...).

**Comment by jhiswin** [comment]
So my use case is somewhat specific and is due to the user specifically requiring it.  Sorry the explanation is a bit long.
I'm working with an app for students to input their answers.  Overview is that it is for younger than high school children.  One of the requirements is for students to be able to input using only keyboard input, as well as still be able to enter plaintext within their input.  Meaning that the command style input for symbols has been set as a user requirement.  
He wants a list of symbols that are activated by "\\", and further requirements were to introduce "\\t" for "\\times" and "\\d" for \\div symbol, which I recommended against due to conflicts with other symbols.  He states that learning latex style input is an educational experience from a pedagogical point of view and also that it is convenient to input.  

I do think that having a standard "escape" style input for symbols is a good thing, as it feels inline and is pretty natural, and lends naturally to the concept of escaped symbols to the user.  I was going to ask about introducing \\nthroot like MathQuill does, because I thought the \\sqrt[n]{x} syntax a bizarre choice (which I understand is due to the typesetting and function implementation aspects of tex).  

Hotkey (alt+v) style input will probably be undesirable.  

What I might do just for my own personal situation is to create a hinter that just shows blank boxes with the latex notation (for example in code hinting style) \\frac{_}{_} which I think MathQuill used in the early days.

I can explain more in detail the situation, but it might devolve into kind of a rant about the project I'm in sort of a consulting position with.
TL;DR something like visual assisted Markdown felt the most natural, similarly command style symbols being like emoticons is desirable
MathQuill's minimalism had him sold.  MathLive's handling and rendering in blocks felt off.

Prior to finally getting him to try a demo with MathQuill which I injected into the app as a POC, his team had devised a poor man's asciimath with latex symbols as the syntax for plaintext input, with a separated equation editor that generates math text (which has now been proven unwieldy).  He agreed with it at first, because he was convinced by them that being able to enter plain text would be more convenient to use and that the text data would be simpler through direct text comparison of answers, until he saw how natural and easy to use MathQuill was and it was proven that latex comparison using a CAS and MathJax rendering, which they ended up having to use anyways, solved any issues there might be.

**Comment by jhiswin** [comment]
TL;DR summary
A user set requirement.  From the point of view of a teacher designing curriculum who is delivering this as a product's input method for students:
- Being able to just use text input is important
- Something like visual assisted Markdown felt the most natural.
- Similarly command style symbols being like emoticons is desirable, and considered to have the added bonus of having educational value.
- Direct replacing normal text without an identifier like a hash tag or escape character is not an option and was also considered unnatural.
- Hotkeys are undesirable as is having a bunch of buttons for symbols (as the only option)


The minimalism of MathQuill, and its way of immediate rendering what you type instead of handling it in blocks, was also desirable.
(To the point that they devised it using a plain text representation of math as input)

For my own use case, I will probably try to implement something like \\frac{\_}{\_} as a hinter, where the user will see \\frac{_}{_} and can just input the two parameters and then have it rendered in a way similar to code hinters in IDEs.

**Comment by arnog** [comment]
Great, thank you for the explanation, that's super useful.

Could you expand on this:
> The minimalism of MathQuill, and its way of immediate rendering what you type instead of handling it in blocks, was also desirable.

I'm not sure what you mean by "handling in blocks"? Could you give me an example?

**Comment by NSoiffer** [comment]
I'm jumping in a bit late, but why not just use the [ASCIIMath ](http://asciimath.org/)syntax that mathlive (mostly) supports? The symbols and meanings are mostly intuitive (e.g. "<=", "sqrt", "/", "^"). In mathlive, they immediately convert, but you can also write it elsewhere and paste it in and it will convert to 2D math. Seems much more natural that teaching kids a bunch of things like \frac.

**Comment by jhiswin** [comment]
So the handling in blocks was something mentioned by the user when he was trying MathLive and he said that it seemed to resolve in blocks while typing. 
The hinting such as the drawing of boxes and the tooltips for commands.  I thought they would be advantages, but he seems to have conflated it with the `^` bug behavior, and described it as contributing to information overload.
\\sqrt not immediately resolving to a symbol after typing is also part of what he is referring to.

The user was also reacting to ASCIIMath syntax not working with MathLive/MathQuill style input in reference to the "handling in blocks" vs "immediate" rendering.

Related, is that I also have to hook the enter key as a signal for answer submission, so using enter to signal the end of a symbol won't work.

@NSoiffer The \\frac example I used was not a good one.  Using `/` is definitely best.  Typing `sqrt` doesn't work, but that is an example where they deviated from ASCIIMath when devising their own input syntax and ended up using latex style commands.  I brought it up, basically that ASCIIMath already exists, and he replied with the need to allow plain text in the input, and asked how I would implement that without conflicts.
What you mention about copy pasting sounds close to what they have implemented, where they have an equation editor that generates syntax that they can copy and paste into the input boxes.  That did not work well in actual use, but perhaps not having MathLive style live rendering on paste was the problem.
Being able to copy paste ASCIIMath is a cool feature.  I'll try it out and see if it works.  Although I'm not sure how well having to copy and paste between boxes will work out.
I see that things like xx and -: work.  Which is great.

So far it seems like `\\name` for symbols is still the simplest and most consistent.
Some things that are currently needed are: \\times(using \\t) \\div(using \\d) \\nthroot(?) \\angle \\overline \\triangle \\le \\lt \gt \ge
The user has said that he would like a simpler way to enter symbols.  xx and -: look like good alternatives for the above, but I don't see a good alternative for all of them.  It would lead to creating a mish-mash of ASCIIMath and latex commands, because conflicts such as `/` and `/_` and `/_\` can't be resolved.
I brought up using buttons, which he said would be good to have if they are designed in a minimal way, and obviously required for the mobile version, but that being able to just input it directly inline is preferrable.

I had originally recommended latex on the basis of it being a standard reliable way to represent equations in data, so that we could be sure that parsing would not be a problem.  Since a simple CAS for answer comparison would inevitably be required.

**Comment by arnog** [comment]
Thanks again for taking the time to writeup this. Very useful.

> So the handling in blocks was something mentioned by the user when he was trying MathLive and he said that it seemed to resolve in blocks while typing.

I see. That's not how MathLive works. It renders the equation as soon as you type. Same as MathQuill, really.

Note also, as Neil mentioned, that MathLive does support MathASCII syntax. FYI, the default MathASCII shortcuts that are supported are here: https://mathlive.io/reference_shortcuts.html 

You can define your own inline shortcuts if desired.

You can also define your own macros (i.e. `\t`).

And as you probably already know, you can also customize the buttons that appear in the virtual keyboard.

That said, and again to agree with Neil, I would find it odd in a pedagogic setting to invent and make students learn a new syntax when at least two already exist (Latex and MathASCII).

That sounds like an interesting project. Best of luck!
